imagemanager/imagemanager.go

```diff
@@ -0,0 +1,993 @@
+	AddService(service ServiceInfo) error
+	SetLayerCached(digest string, cached bool) error
+	SetServiceCached(serviceID string, cached bool) error
+	RemoveService(serviceID string, aosVersion uint64) error
+	RemoveLayer(digest string) error
+}
+
+// CryptoContext interface to access crypto functions.
+type CryptoContext interface {
+	ImportSessionKey(keyInfo fcrypt.CryptoSessionKeyInfo) (fcrypt.SymmetricContextInterface, error)
+	CreateSignContext() (fcrypt.SignContextInterface, error)
+}
+
+// Imagemanager image manager instance.
+type Imagemanager struct {
+	layersDir              string
+	servicesDir            string
+	tmpDir                 string
+	storage                Storage
+	cryptoContext          CryptoContext
+	serviceAllocator       spaceallocator.Allocator
+	layerAllocator         spaceallocator.Allocator
+	tmpAllocator           spaceallocator.Allocator
+	gidPool                *uidgidpool.IdentifierPool
+	serviceTTLDays         uint64
+	layerTTLDays           uint64
+	validateTTLStopChannel chan struct{}
+	fileServer             *fileserver.FileServer
+}
+
+// ServiceInfo service information.
+type ServiceInfo struct {
+	aostypes.ServiceInfo
+	RemoteURL string
+	Path      string
+	Timestamp time.Time
+	Cached    bool
+	GID       int
+	Config    aostypes.ServiceConfig
+	Layers    []string
+}
+
+// LayerInfo service information.
+type LayerInfo struct {
+	aostypes.LayerInfo
+	RemoteURL string
+	Path      string
+	Timestamp time.Time
+	Cached    bool
+}
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+var (
+	// ErrNotExist not exist service error.
+	ErrNotExist = errors.New("service not exist")
+
+	// ErrVersionMismatch new service version <= existing one.
+	ErrVersionMismatch = errors.New("version mismatch")
+
+	// NewSpaceAllocator space allocator constructor.
+	// nolint:gochecknoglobals // used for unit test mock
+	NewSpaceAllocator = spaceallocator.New
+)
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+// New creates new image manager object.
+func New(cfg *config.Config, storage Storage, cryptoContext CryptoContext) (imagemanager *Imagemanager, err error) {
+	imagemanager = &Imagemanager{
+		layersDir:              path.Join(cfg.ImageStoreDir, "layers"),
+		servicesDir:            path.Join(cfg.ImageStoreDir, "services"),
+		tmpDir:                 path.Join(cfg.ImageStoreDir, "tmp"),
+		storage:                storage,
+		cryptoContext:          cryptoContext,
+		serviceTTLDays:         cfg.ServiceTTLDays,
+		layerTTLDays:           cfg.LayerTTLDays,
+		gidPool:                uidgidpool.NewGroupIDPool(),
```

Comment:
on the startup you have to all all gids which are already in use

Reply:
Done